### PR TITLE
Add LEP for feature Extend CSI Snapshot support to Longhorn snapshot

### PR DESCRIPTION
**Pull request by PhanLe1010**
longhorn/longhorn#2534

### CODE REVIEW

File: enhancements/20220110-extend-csi-snapshot-to-support-longhorn-snapshot.md [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+### API changes
+None
+
+## Design
+
+### Implementation Overview
+
+We follow the specification in [the CSI spec](https://github.com/container-storage-interface/spec/blob/master/spec.md#createsnapshot) when supporting the CSI snapshot.
+
+We define a new parameter in the VolumeSnapshotClass `type`.
```
Comment (shuo-wu) [inline enhancements/20220110-extend-csi-snapshot-to-support-longhorn-snapshot.md:96]: What if this field is not set? Based on the below test plan, it seems that Longhorn will fall back to create a backup.

File: enhancements/20220110-extend-csi-snapshot-to-support-longhorn-snapshot.md [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+  We cannot take a snapshot of non-attached volume.
+* Check if a Longhorn snapshot with the same name as the requested CSI snapshot already exists. 
+  If yes, return OK without taking a new Longhorn snapshot.
+* Take a new Longhorn snapshot. Encode the snapshotId in the format `snap://volume-name/snapshot-name`. 
+  This snaphotId will be used in the later CSI CreateVolume and DeleteSnapshot call. 
+  
+In [CreateVolume function](https://github.com/longhorn/longhorn-manager/blob/878cfb868c568396d6ebfa4ce096c5d95d9b31e3/csi/controller_server.go#L63):
+* If the VolumeContentSource is a `VolumeContentSource_Snapshot` type, decode the snapshotId in the format from the above step. 
+* Create a new [Longhorn VolumeDatasource](https://github.com/longhorn/longhorn-manager/blob/878cfb868c568396d6ebfa4ce096c5d95d9b31e3/types/data_source.go#L20)
+  object from the source volume and snapshot name and set it as the `dataSource` of the new Longhorn volume.
```
Comment (shuo-wu) [inline enhancements/20220110-extend-csi-snapshot-to-support-longhorn-snapshot.md:115]: This should be similar to the volume clone feature, right? 

This statement `Create a new Longhorn VolumeDatasource object` is a little bit confusing. Longhorn does not contain a CRD named `VolumeDatasource`, right? Instead, the CSI plugin just asks the manager to create a new Volume CR but with field `spec.dataSource` set.

**Review by shuo-wu — COMMENTED** [review]
In general LGTM

File: enhancements/20220110-extend-csi-snapshot-to-support-longhorn-snapshot.md (hunk 1 above)
Comment (PhanLe1010) [reply enhancements/20220110-extend-csi-snapshot-to-support-longhorn-snapshot.md:96]: Yeah, for backward capacity, empty `type` mean `longhorn-backup`

File: enhancements/20220110-extend-csi-snapshot-to-support-longhorn-snapshot.md (hunk 2 above)
Comment (PhanLe1010) [reply enhancements/20220110-extend-csi-snapshot-to-support-longhorn-snapshot.md:115]: Yeah, you are right, there is no VolumeDatasource CRD. I am referring to the in-memory object https://github.com/longhorn/longhorn-manager/blob/9e4229d62d1dd591b43d7183bfbd0b403cfdc454/csi/controller_server.go#L97-L99

The manager to create a new Volume CR but with field spec.dataSource set

Comment (shuo-wu) [reply enhancements/20220110-extend-csi-snapshot-to-support-longhorn-snapshot.md:115]: OK. It's better to clarify it in the doc.

**Review by innobead — DISMISSED** [review]
LGTM

File: enhancements/20220110-extend-csi-snapshot-to-support-longhorn-snapshot.md [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+The value of the parameter `type` can be `longhorn-snapshot` or `longhorn-backup`.
+When `type` is `longhorn-snapshot` it means that the CSI VolumeSnapshot created with this VolumeSnapshotClass is associated with a Longhorn snapshot.
+When `type` is `longhorn-backup` it means that the CSI VolumeSnapshot created with this VolumeSnapshotClass is associated with a Longhorn backup.
+
+In [CreateSnapshot function](https://github.com/longhorn/longhorn-manager/blob/878cfb868c568396d6ebfa4ce096c5d95d9b31e3/csi/controller_server.go#L539), we get the
+value of parameter `type`. If it is `longhorn-backup`, we take a Longhorn backup as before. If it is `longhorn-snapshot` we do:
+* Get the name of the Longhorn volume
+* Check if the volume is in attached state.
+  If it is not, return `codes.FailedPrecondition`. 
+  We cannot take a snapshot of non-attached volume.
```
[the remark this answers is not in the document]
Comment (PhanLe1010) [reply enhancements/20220110-extend-csi-snapshot-to-support-longhorn-snapshot.md:106]: Thanks David

File: enhancements/20220110-extend-csi-snapshot-to-support-longhorn-snapshot.md (hunk 2 above)
Comment (PhanLe1010) [reply enhancements/20220110-extend-csi-snapshot-to-support-longhorn-snapshot.md]: Updated. Thank you

File: enhancements/20220110-extend-csi-snapshot-to-support-longhorn-snapshot.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (innobead) [reply enhancements/20220110-extend-csi-snapshot-to-support-longhorn-snapshot.md:106]: cc @c3y1huang

**Review by innobead — APPROVED** [review]

**Pull request merged**